File: XXO/XXO/XXOGame.m
//
//  XXOGame.m
//  XXO
//
//  Created by Paul Jacobs on 1/19/15.
//  Copyright (c) 2015 Paul Jacobs. All rights reserved.
//

#import "XXOGame.h"

#pragma mark -

@implementation XXOGame

#pragma mark Setup
- (instancetype)initWithDelegate:(id<XXOGameDelegate>)del
{
    self = [super init];
    if (self) {
        self.delegate = del;
        // if game was interrupted, load from storage.
        [self resetGame];
    }
    return self;
}


#pragma mark Game Actions
- (player)playAtSpace:(boardSpace)spaceNumber
{
    if (spaceNumber >= 0 && spaceNumber <= 8) {
        if (self.currentPlayer == playerX) {
            self.board[spaceNumber] = [NSNumber numberWithInt:playerX];

            self.currentPlayer = playerO;
            [self.delegate player:playerX didPlayAtSpace:spaceNumber];
        } else if (self.currentPlayer == playerO) {
            self.board[spaceNumber] = [NSNumber numberWithInt:playerO];
            self.currentPlayer = playerX;
            [self.delegate player:playerO didPlayAtSpace:spaceNumber];
        }
    }
    
    player winningPlayer = [self checkForWin];
    if (winningPlayer != blank) {
        [self.delegate gameOverWithWinner:winningPlayer];
    }
    
    return winningPlayer;
}

- (player)checkForWin
{
    player winningPlayer = blank;
 
    if ([self.board[upperLeft] isEqualToNumber: self.board[upperMid]] &&
        [self.board[upperLeft] isEqualToNumber:self.board[upperRight]] &&
        ![self.board[upperLeft] isEqualToNumber:[NSNumber numberWithInt:blank]]) {
        // Upper Row Winner
        winningPlayer = (player)[self.board[upperLeft] integerValue];
    } else
    if ([self.board[centerLeft] isEqualToNumber: self.board[centerMid]] &&
        [self.board[centerLeft] isEqualToNumber:self.board[centerRight]] &&
        ![self.board[centerLeft] isEqualToNumber:[NSNumber numberWithInt:blank]]) {
        // Center Row Winner
        winningPlayer = (player)[self.board[centerLeft] integerValue];
    } else
    if ([self.board[lowerLeft] isEqualToNumber: self.board[lowerMid]] &&
        [self.board[lowerLeft] isEqualToNumber:self.board[lowerRight]] &&
        ![self.board[lowerLeft] isEqualToNumber:[NSNumber numberWithInt:blank]]) {
        //  Lower Row Winner
        winningPlayer = (player)[self.board[lowerLeft] integerValue];
    } else

    if ([self.board[upperLeft] isEqualToNumber: self.board[lowerLeft]] &&
        [self.board[upperLeft] isEqualToNumber:self.board[centerLeft]] &&
        ![self.board[upperLeft] isEqualToNumber:[NSNumber numberWithInt:blank]]) {
        // Left Col Winner
        winningPlayer = (player)[self.board[upperLeft] integerValue];
    } else
    if ([self.board[upperMid] isEqualToNumber: self.board[centerMid]] &&
        [self.board[upperMid] isEqualToNumber:self.board[lowerMid]] &&
        ![self.board[upperMid] isEqualToNumber:[NSNumber numberWithInt:blank]]) {
        // Middle Col Winner
        winningPlayer = (player)[self.board[upperMid] integerValue];
    } else
    if ([self.board[upperRight] isEqualToNumber: self.board[lowerRight]] &&
        [self.board[upperRight] isEqualToNumber:self.board[centerRight]] &&
        ![self.board[upperRight] isEqualToNumber:[NSNumber numberWithInt:blank]]) {
        // Right Col Winner
        winningPlayer = (player)[self.board[upperRight] integerValue];
    } else

    if ([self.board[upperLeft] isEqualToNumber: self.board[lowerRight]] &&
        [self.board[upperLeft] isEqualToNumber:self.board[centerMid]] &&
        ![self.board[upperLeft] isEqualToNumber:[NSNumber numberWithInt:blank]]) {
        // Diagonal Upper Left <-> Lower Right Winner
        winningPlayer = (player)[self.board[upperLeft] integerValue];
    } else
    if ([self.board[upperRight] isEqualToNumber: self.board[lowerLeft]] &&
        [self.board[upperRight] isEqualToNumber:self.board[centerMid]] &&
        ![self.board[upperRight] isEqualToNumber:[NSNumber numberWithInt:blank]]) {
        // Diagonal Upper Right <-> Lower Left Winner
        winningPlayer = (player)[self.board[upperRight] integerValue];
    }

    return winningPlayer;
}

- (void)resetGame
{
    self.board = [@[] mutableCopy];
    for (int i = 0; i<9;i++) {
        [((NSMutableArray*)self.board) addObject:[NSNumber numberWithInt:blank]];
    }
    self.currentPlayer = (arc4random()%2==0 ? playerO : playerX);
    if (self.delegate) {
        [self.delegate gameDidReset];
    }
}

- (void)saveGame
{
    // Save game data to app storage

}

- (void)loadGame
{
    // Load game data from app storage
    
    if (self.delegate) {
        [self.delegate gameDidLoad];
    }
}

- (void)encodeWithCoder:(NSCoder *)coder {
    [coder encodeObject:self.board forKey:@"board"];
    [coder encodeInteger:self.currentPlayer forKey:@"currentPlayer"];
}

/*
-(NSArray*)board
{
    return [self.board copy];
}
*/

@end

File: XXO/XXO/XXOGame.swift
//
//  XXOGame.swift
//  XXO
//
//  Created by Paul Jacobs on 1/23/15.
//  Copyright (c) 2015 Paul Jacobs. All rights reserved.
//

import Foundation

enum player : Int {
    case blank = 0
    case playerX = 1
    case playerO = 2
    case Tie = 3
}

enum boardSpace : Int {
    case upperLeft  = 0
    case upperMid   = 1
    case upperRight = 2
    case centerLeft = 3
    case centerMid  = 4
    case centerRight = 5
    case lowerLeft  = 6
    case lowerMid   = 7
    case lowerRight = 8
}

protocol XXOGameDelegate {
    func playerDidPlayAtSpace(p:player, space:boardSpace)
    func gameDidReset()
    func gameDidLoad()
    func gameOverWithWinner(p:player)
}

@objc class XXOGame
{
    var board : Array<player> = []
    var currentPlayer : player = .blank
    var delegate : XXOGameDelegate
    
    init(d:XXOGameDelegate)
    {
        delegate = d
        resetGame()
    }
    
    func playAtSpace(space:boardSpace) -> player
    {
        if (space.rawValue <= 8 &&
            board[space.rawValue] == .blank &&
            currentPlayer != .blank) {
            
            switch (currentPlayer) {
                case .playerX, .playerO:
                    board[space.rawValue] = currentPlayer;
                    currentPlayer = (currentPlayer == .playerX ? .playerO : .playerX);
                    delegate.playerDidPlayAtSpace(currentPlayer, space: space)
            default:
                // ignore attempts to play when game is not being played
                break
            }
        }
        
        let winningPlayer = checkForWin()
        if (winningPlayer != .blank) {
            delegate.gameOverWithWinner(winningPlayer)
        }
        saveGame()
        return winningPlayer;
    }
    
    
    
    func checkForWin() -> player
    {
        var winningPlayer = player.blank
        
        if (board[boardSpace.upperLeft.rawValue] == board[boardSpace.upperMid.rawValue] &&
            board[boardSpace.upperLeft.rawValue] == board[boardSpace.upperRight.rawValue] &&
            board[boardSpace.upperLeft.rawValue] != player.blank)
        {   // Upper row winner
            winningPlayer = self.board[boardSpace.upperLeft.rawValue]
        } else
        if (board[boardSpace.centerLeft.rawValue] == board[boardSpace.centerMid.rawValue] &&
            board[boardSpace.centerLeft.rawValue] == board[boardSpace.centerRight.rawValue] &&
            board[boardSpace.centerLeft.rawValue] != player.blank)
        {   // Center row winner
            winningPlayer = self.board[boardSpace.centerLeft.rawValue]
        } else
        if (board[boardSpace.lowerLeft.rawValue] == board[boardSpace.lowerMid.rawValue] &&
            board[boardSpace.lowerLeft.rawValue] == board[boardSpace.lowerRight.rawValue] &&
            board[boardSpace.lowerLeft.rawValue] != player.blank)
        {   // Lower row winner
            winningPlayer = self.board[boardSpace.lowerLeft.rawValue]
        } else
        if (board[boardSpace.upperLeft.rawValue] == board[boardSpace.lowerLeft.rawValue] &&
            board[boardSpace.upperLeft.rawValue] == board[boardSpace.centerLeft.rawValue] &&
            board[boardSpace.upperLeft.rawValue] != player.blank)
        {   // Left Col winner
            winningPlayer = self.board[boardSpace.upperLeft.rawValue]
        } else
        if (board[boardSpace.upperMid.rawValue] == board[boardSpace.centerMid.rawValue] &&
            board[boardSpace.upperMid.rawValue] == board[boardSpace.lowerMid.rawValue] &&
            board[boardSpace.upperMid.rawValue] != player.blank)
        {   // Middle Col winner
            winningPlayer = self.board[boardSpace.upperMid.rawValue]
        } else
        if (board[boardSpace.upperRight.rawValue] == board[boardSpace.lowerRight.rawValue] &&
            board[boardSpace.upperRight.rawValue] == board[boardSpace.centerRight.rawValue] &&
            board[boardSpace.upperRight.rawValue] != player.blank)
        {   // Right Col winner
            winningPlayer = self.board[boardSpace.upperRight.rawValue]
        } else
        if (board[boardSpace.upperLeft.rawValue] == board[boardSpace.lowerRight.rawValue] &&
            board[boardSpace.upperLeft.rawValue] == board[boardSpace.centerMid.rawValue] &&
            board[boardSpace.upperLeft.rawValue] != player.blank)
        {   // Diagonal Upper Left <-> Lower Right Winner
            winningPlayer = self.board[boardSpace.upperLeft.rawValue]
        } else
        if (board[boardSpace.upperRight.rawValue] == board[boardSpace.lowerLeft.rawValue] &&
            board[boardSpace.upperRight.rawValue] == board[boardSpace.centerMid.rawValue] &&
            board[boardSpace.upperRight.rawValue] != player.blank)
        {   // Diagonal Upper Right <-> Lower Left Winner
            winningPlayer = self.board[boardSpace.upperRight.rawValue]
        } else if (( // had to break up this expression because swift can't handle it as one piece yet.
            board[0] != player.blank &&
            board[1] != player.blank ) && (
            board[2] != player.blank &&
            board[3] != player.blank ) && (
            board[4] != player.blank &&
            board[5] != player.blank ))
        {
            if (board[6] != player.blank &&
                board[7] != player.blank &&
                board[8] != player.blank )
            { // Tie
                winningPlayer = player.Tie
            }
        } else {
            winningPlayer = player.blank
        }
        
        return winningPlayer
    }
    
    func resetGame()
    {
        board = [   .blank, .blank, .blank,
                    .blank, .blank, .blank,
                    .blank, .blank, .blank, ];
        currentPlayer = ( arc4random()%2==0 ? player.playerO : player.playerX);
        delegate.gameDidReset()
        saveGame()
    }
    
    func saveGame()
    {
        var saveBoard : [Int] = []
        let def = NSUserDefaults.standardUserDefaults()

        for index in 0..<9 {
            saveBoard.append(board[index].rawValue)
        }
        
        def.setObject(saveBoard, forKey: "board")
        def.setInteger(currentPlayer.rawValue, forKey: "currentPlayer")
        def.synchronize()
    }
    
    func loadGame()
    {
        var pre_board : [Int] = []
        let def = NSUserDefaults.standardUserDefaults()
        pre_board = def.objectForKey("board") as [Int]
        if (pre_board.count == 9) {
            for index in 0..<9 {
                board[index] = player(rawValue: index)!
            }
            currentPlayer = player(rawValue: def.integerForKey("currentPlayer"))!
        } else {
            resetGame()
        }
        delegate.gameDidLoad()
    }
    
}